tests/rados/test_bilog_trim.py

```diff
@@ -0,0 +1,143 @@
+
+Test needs to run as part of rgw/tier-2_rgw_rados_multisite_ecpool.yaml
+as it needs a rgw multisite setup along with data present in RGW pools
+"""
+
+import datetime
+import random
+import time
+from math import floor
+
+from ceph.ceph_admin import CephAdmin
+from ceph.rados import utils as rados_utils
+from ceph.rados.core_workflows import RadosOrchestrator
+from utility.log import Log
+
+log = Log(__name__)
+
+
+def run(ceph_cluster, **kw):
+    """
+    #CEPH-83575437
+    #BZ-2056818
+    This test is to verify no issues are observed during deep-scrub when rgw bilog
+    trimming occurs with OSD down
+    Steps:
+    Ref:
+    - https://bugzilla.redhat.com/show_bug.cgi?id=2056818#c42
+    - https://bugzilla.redhat.com/show_bug.cgi?id=2056818#c50
+    1. Deploy minimal RGW multisite setup
+    2. Add data to both the sites in S3 buckets and let the minimal client IO running
+    3. Keep running the deep-scrub for index pool PGs
+    4. Set noout flag
+    5. stop the secondary OSD on the primary site(site1)
+    6. run the following command
+       radosgw-admin bilog autotrim
+    7. Once the above command finishes, start the primary OSD that was stopped in step 4
+    8. Retrigger the deep-scrub for eight PGs in index pool
+    """
+    log.info(run.__doc__)
+    config = kw["config"]
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    rados_obj = RadosOrchestrator(node=cephadm)
+
+    try:
+        # set noout flag
+        if not rados_utils.configure_osd_flag(ceph_cluster, action="set", flag="noout"):
+            log.error("Could not set noout flag on the cluster")
+            raise Exception("Could not set noout flag on the cluster")
+
+        # get acting set for the rgw index pool
+        _pool_name = "primary.rgw.buckets.index"
+        acting_set = rados_obj.get_pg_acting_set(pool_name=_pool_name)
+        log.info(f"Acting set for pool {_pool_name}: {acting_set}")
+
+        # secondary osd of 1st PG from rgw index pool
+        second_osd = acting_set[1]
+
+        # get list on active OSDs
+        osd_list = rados_obj.get_osd_list(status="up")
+
+        # bring down 30% of total UP OSDs
+        osd_down = random.choices(osd_list, k=floor(0.3 * len(osd_list)))
+        osd_down.append(second_osd)
+
+        log.info(f"List of OSD to be stopped: {osd_down}")
+
+        # trigger deep-scrub on the rgw index pool
+        rados_obj.run_deep_scrub(pool=_pool_name)
+
+        for osd_id in osd_down:
+            if not rados_obj.change_osd_state(action="stop", target=int(osd_id)):
+                log.error(f"Could not stop OSD.{osd_id}")
+                raise Exception(f"Could not stop OSD.{osd_id}")
+
+        # for a duration of 7 mins, trigger bi-log trimming along with deep-scrub
+        timeout_time = datetime.datetime.now() + datetime.timedelta(seconds=420)
+        while datetime.datetime.now() < timeout_time:
+            cephadm.shell(args=["radosgw-admin bilog autotrim"])
```

Comment:
Any way that we can confirm that bilog trimming was actually triggered and completed in the 7 min timeframe provided?

Reply:
There is a way to confirm if we run bilog trimming after writing IOs, but now I have added code to run background IOs, with IOs in place, the list of bilog to be trimmed gets updated and therefore no way to confirm that bilog trimming got completed while IOs are still running because it now becomes a continuous cycle